src/hotspot/share/opto/vectornode.hpp

```diff
@@ -778,6 +778,63 @@ class StoreVectorNode : public StoreNode {
                                                      idx == MemNode::ValueIn + 1; }
 };
 
+class StoreVectorMaskedNode : public StoreVectorNode {
+ public:
+  StoreVectorMaskedNode(Node* c, Node* mem, Node* dst, Node* src, const TypePtr* at, Node* mask)
+   : StoreVectorNode(c, mem, dst, at, src) {
+    assert(mask->bottom_type()->is_long(), "sanity");
+    init_class_id(Class_StoreVector);
+    set_mismatched_access();
+    add_req(mask);
+  }
+
+  virtual int Opcode() const;
+
+  virtual uint match_edge(uint idx) const {
+    return idx > 1;
+  }
+  Node* Ideal(PhaseGVN* phase, bool can_reshape);
+
+  static StoreVectorMaskedNode* make(int opc, Node* ctl, Node* mem, Node* dst, Node* src,
+                                    const TypePtr* atype, Node* mask);
+};
+
+class LoadVectorMaskedNode : public LoadVectorNode {
+ public:
+  LoadVectorMaskedNode(Node* c, Node* mem, Node* src, const TypePtr* at, const TypeVect* vt, Node* mask)
+   : LoadVectorNode(c, mem, src, at, vt) {
+    assert(mask->bottom_type()->is_long(), "sanity");
+    init_class_id(Class_LoadVector);
+    set_mismatched_access();
+    add_req(mask);
+  }
+
+  virtual int Opcode() const;
+
+  virtual uint match_edge(uint idx) const {
+    return idx > 1;
+  }
+  Node* Ideal(PhaseGVN* phase, bool can_reshape);
+
+  static LoadVectorMaskedNode* make(int opc, Node* ctl, Node* mem, Node* src,
+                                const TypePtr* atype, const TypeVect* vt,
+                                Node* mask);
+};
+
+class VectorMaskGenNode : public TypeNode {
+ public:
+  VectorMaskGenNode(Node* src, const Type* ty, const Type* ety): TypeNode(ty, 2), _elemType(ety) {
+    init_req(1, src);
+  }
+
+  virtual int Opcode() const;
+  const Type* get_elem_type()  { return _elemType;}
+
+  static VectorMaskGenNode* make(int opc, Node* src, const Type* ty, const Type* ety);
+  private:
+   const Type* _elemType;
```

Comment:
As currently there is no support for mask registers in RA, for X86 long ideal type is sufficient for a mask producing node (def operand is a mask register) ;  But for complete support returning Op_RegVMask as an ideal_reg() type for masked Ideal node should do the trick without creating an explicit new ideal Type for mask generating nodes.  Spill sizes and number of slots may be different for X86 and ARM (SVE).  
Shallow copy during Node::clone should be sufficient here since encapsulated element type will be preserved.

Reply:
> As currently there is no support for mask registers in RA, for X86 long ideal type is sufficient for a mask producing node (def operand is a mask register) ; But for complete support returning Op_RegVMask as an ideal_reg() type for masked Ideal node should do the trick without creating an explicit new ideal Type for mask generating nodes. Spill sizes and number of slots may be different for X86 and ARM (SVE).

So, do you have a plan to support Op_RegVMask? In SVE, we will use this kind of node for mask/predicate type.

> Shallow copy during Node::clone should be sufficient here since encapsulated element type will be preserved.

Checking the code in Node::clone() again, I think the object copy relies on size_of(), so you need to override that to get the correct object size for copying.

Reply:
> > As currently there is no support for mask registers in RA, for X86 long ideal type is sufficient for a mask producing node (def operand is a mask register) ; But for complete support returning Op_RegVMask as an ideal_reg() type for masked Ideal node should do the trick without creating an explicit new ideal Type for mask generating nodes. Spill sizes and number of slots may be different for X86 and ARM (SVE).
> 
> So, do you have a plan to support Op_RegVMask? In SVE, we will use this kind of node for mask/predicate type.
> 
Not as the part of this patch but as a separate RFE,  we may benefit from decoupling b/w ideal type(bottom_type) and ideal_reg() for a given Ideal node; this should allow us to build any future extension on top of these masked generating nodes (VectorMaskGen).

> > Shallow copy during Node::clone should be sufficient here since encapsulated element type will be preserved.
> 
> Checking the code in Node::clone() again, I think the object copy relies on size_of(), so you need to override that to get the correct object size for copying.

Thanks for pointing out;  I missed this earlier.